src/server/frontend_wayland/ext_image_capture_v1.cpp

```diff
@@ -0,0 +1,607 @@
+{
+    if (capture_has_been_called)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::already_captured,
+            "Cannot attach buffer after capture"));
+    }
+
+    auto buf = wayland::Buffer::from(buffer);
+    if (!buf)
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error(
+            "wl_buffer@" +
+            std::to_string(wl_resource_get_id(buffer)) +
+            " is not a wayland::Buffer"));
+    }
+    target = wayland::make_weak(buf);
+}
+
+void mf::ExtImageCopyCaptureFrameV1::damage_buffer(
+    int32_t x,
+    int32_t y,
+    int32_t width,
+    int32_t height)
+{
+    if (!session)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::invalid_buffer_damage,
+            "session is gone"));
+    }
+    // TODO: this needs to be converted to output space coordinates
+    session.value().apply_damage(geom::Rectangle{{x, y}, {width, height}});
+}
+
+void mf::ExtImageCopyCaptureFrameV1::capture()
+{
+    if (capture_has_been_called)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::already_captured,
+            "capture already called"));
+    }
+    if (!target)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::no_buffer,
+            "no buffer attached"));
+    }
+    capture_has_been_called = true;
+
+    if (!session)
+    {
+        send_failed_event(FailureReason::stopped);
+        return;
+    }
+    session.value().maybe_capture_frame();
+}
+
+bool mf::ExtImageCopyCaptureFrameV1::is_ready() const
+{
+    return capture_has_been_called;
+}
+
+void mf::ExtImageCopyCaptureFrameV1::capture_frame(
+    geom::Rectangle output_space_area,
+    glm::mat2 transform,
+    bool overlay_cursor,
+    geom::Size buffer_size,
+    geom::Rectangle buffer_space_damage)
+{
+    if (!session)
+    {
+        send_failed_event(FailureReason::stopped);
+        return;
+    }
+
+    auto shm_buffer = dynamic_cast<ShmBuffer*>(wayland::as_nullable_ptr(target));
+    if (!shm_buffer)
+    {
+        send_failed_event(FailureReason::buffer_constraints);
+        return;
+    }
+    auto shm_data = shm_buffer->data();
+    if (shm_data->format() != mir_pixel_format_argb_8888 ||
+        shm_data->size() != buffer_size)
+    {
+        send_failed_event(FailureReason::buffer_constraints);
+        return;
+    }
+
+    auto buffer = std::shared_ptr<mir::renderer::software::WriteMappable>{
+        shm_data.get(),
+        [shm_data, weak_buffer = target, executor = ctx->wayland_executor](auto*)
+        {
+            // Send release event for underlying buffer if necessary
+            executor->spawn(
+                [shm_data, weak_buffer]()
+                {
+                    if (weak_buffer)
+                    {
+                        weak_buffer.value().send_release_event();
+                    }
+                });
+        }
```

Comment:
We actually don't need to (and shouldn't²) send a `wl_buffer.release` event here; that is only for `wl_surface.attach`¹. The client knows when the buffer is ready for reuse from the `ready` event we send.

¹: And it's actually regarded as somewhat of a mistake there, too 🤷‍♀️ 
²: The [protocol](https://wayland.app/protocols/ext-image-copy-capture-v1#ext_image_copy_capture_frame_v1:request:attach_buffer) here says “the `wl_buffer.release` request (sic) is unused”